apps/designer/app/designer/features/style-panel/controls/text/text-control.tsx

```diff
@@ -15,37 +39,195 @@ export const TextControl = ({
     property: styleConfig.property,
   });
 
+  const inputRef = useRef<HTMLInputElement | null>(null);
+
+  if (value === undefined) return null;
+
   const setValue = setProperty(styleConfig.property);
 
+  const items = styleConfig.items.map(({ label }) => label);
+  const units = sortedUnits.map((unit) => unit);
   const handleChange = (
-    styleValue: StyleValue | undefined,
+    type: string,
+    item: string | number | undefined,
     isEphemeral: boolean
   ) => {
-    if (styleValue === undefined) {
-      setValue("");
-      return;
-    }
-    const { value, type } = styleValue;
+    if (!item) return;
     const newValue =
-      type === "unit" ? `${value}${styleValue.unit}` : `${value}`;
-    setValue(newValue, { isEphemeral });
+      type === "unit" ? value.value + String(item) : String(item);
+    setValue(value.type === "unit" ? newValue + value.unit : newValue, {
+      isEphemeral,
+    });
+  };
+  const handleKeyDown = (event: KeyboardEvent<HTMLInputElement>) => {
+    handleKeyDownEnter(event);
+    if (event.currentTarget.matches("[aria-activedescendant]")) return true;
+    if (value.type !== "unit") return true;
+    if (!["ArrowUp", "ArrowDown"].includes(event.code)) return false;
+    event.preventDefault();
+    handleKeyDownArrowUpDown(event);
+  };
+  const handleKeyDownEnter = (event: KeyboardEvent<HTMLInputElement>) => {
+    if (
+      event.code === "Enter" &&
+      String(value.value) !== event.currentTarget.value
+    ) {
+      setValue(event.currentTarget.value);
+      const number = parseFloat(event.currentTarget.value);
+      if (!isNaN(number)) event.currentTarget.value = String(number);
+    }
+  };
+  const handleKeyDownArrowUpDown = (event: KeyboardEvent<HTMLInputElement>) => {
+    let currentValue = parseFloat(event.currentTarget.value);
+    let currentDelta = 1;
+    if (event.shiftKey) currentDelta = 10;
+    if (event.altKey) currentDelta = 0.1;
+    if (event.code === "ArrowUp") currentValue = currentValue + currentDelta;
+    if (event.code === "ArrowDown") currentValue = currentValue - currentDelta;
+    const currentValueAsString =
+      currentValue % 1
+        ? currentValue.toPrecision(
+            Math.abs(currentValue).toString().indexOf(".") + 2
+          )
+        : String(currentValue);
+    event.currentTarget.value = currentValueAsString;
+    setValue(currentValueAsString, { isEphemeral: true });
   };
 
   return (
-    <CssPropertyControl
-      property={styleConfig.property}
-      value={value}
-      allowedValues={styleConfig.items.map((item) => ({
-        type: "keyword",
-        value: item.name,
-      }))}
-      onChange={(value) => {
-        console.log("change", value);
-        handleChange(value, true);
+    <Combobox
+      name={styleConfig.property}
+      items={items}
+      value={String(value.value)}
```

Comment:
we discsussed that combobox should receive the value object, not the value.value string, why is it back?

Reply:
of course it doesn't, you have to pass to the input the value.value

Reply:
I think you must define `itemToString` to make it work. This will also fix filtering.

Reply:
I think I'll make it a required prop so it removes this confusion.

Reply:
\when rendering the value in the input, he can just pass value.value, its in his control where TextField is getting rendered

Reply:
I'm confused. What is "that"? It was working in my branch. Maybe take a look there? Just a heads up: I'm working on some combo box changes in the original branch to 1. Make custom usages easier and 2. To revert the input and prefix behavior I introduced earlier.